### ISSUE TITLE
Customizable Wagner?

### ISSUE THREAD

**Issue by TJung1968** [d3/d3-geo-projection#139]
Recently, I came up with a d3 implementation of the "Customizable Wagner" (see links below).
What is it?
When Wagner presented the equal-area projection that nowadays is called "Wagner VII" (or "Hammer-Wagner") in 1941, he actually didn't present a single projection, but a specific example of his transformation method ("Umbeziffern") that lets you customize the resulting map in regards of length of the pole line, curvature of the parallels and the aspect ratio of the main axes. In 1949, he extended this approach by adding the option to set a controlled areal inflation, showing the "Wagner VIII" (having an inflation of 1.2 at 60° N/S) as example.

So, by passing four parameters to the function, the "Customizable Wagner" allows to render all possible instances of Wagner’s projection.
Here’s an interactive demo which will let you see the results:
https://map-projections.net/d3-customizable-wagner/interactive-example
(The red bordered box "Customize Projection" provides the four parameters, the other two boxes to customize map and map position are merely added for convenience.)

The "static examples" No. 1 to 3 <https://map-projections.net/d3-customizable-wagner/> each show a specific parametrization, plus the full source code.
Static example No. 4 does the same but uses a different approach.
The examples also link to documents providing more information about Wagner’s "Umbeziffern".

However, I do not know whether my code complies with d3 coding standards and frankly, I’m not sure how to contribute.
So if anyone’s interested to add the "Customizable Wagner" to d3-geo-projections: Just grab the code from the links above, feel free to modify it in any way and add it.

**Comment by TJung1968** [comment]
Thank you for your reply!

> wagner7 is the same code as yours

No surprise there, because the d3 wagner7 code is the one that I used to get started. ;-)


>  i.e. .parallels() rather that .lambda1()

Good idea!
I chose the names lambda1 etc. because Wagner used them in his formula, but this way, I guess they’ll be easier to understand.
I think the other names might then be:
psi1() => poleline()
s60() => inflation()
p() => ratio()

I can do that within the next few days, but then...
> Coding style is not difficult to adapt (I will do it if you like)

... I’d be glad if you’d do that!


> seems like you chose wagner7 as default

Hmmm, yes, it’s probably better to choose Wagner VIII as default instead.


> Böhm notation or direct values

In that case, I’d strongly prefer the Böhm notation, because that’s something my mind can handle:
For example, when I want a stronger curvature of the parallels, I can think: "Increase the value from 60 to 70" or something like that. But, in contrast, on the direct values it’s: "Oh, decrease cx() from 2.667233 to 2.305563, cy() from 1.241036 to 1.230614, and increase p() from 0.333333 to 0.388889..." 

Thanks again, I’ll post again when I’ve set the names straight.

**Comment by TJung1968** [comment]
Hello again,

a modified version is now available at
https://map-projections.net/d3-customizable-wagner/

– I renamed the parameters as mentioned above.
– Default values now render the Wagner VIII. I chose this as default, because
a) it’s an established projection (e.g. it was mentioned by Snyder in "Flattening the Earth" and by Canters in "Small-scale Map Projection Design"); and
b) contrary to the Wagner VII, it uses the option to introduce the controlled areal inflation.

For the moment, the option to use the direct values is still available.
Let my tell you why I added this option in the first place: While I’m very interested in map projections, I’m no professional cartographer or something like that. For people like me, I think (as I’ve said) the Böhm notation is much better to handle. However I thought that professionals might prefer the option to use the values that are actually passed to Wagner’s final formula, so I added this option.

Maybe my assumption was wrong, because I guess that those "direct" values are a bit of a problem even for professionals. For example, if you want an equal-area projection, all you have to do using the Böhm notation is to set `inflation(0)`. It doesn’t matter what you note for the other parameters, the result will be equivalent.
But using the direct values, the rule is `m2` must be `1` and `cx * cy * m1 * n` must equal `1`.
Moreover, it’s nigh impossible to validate the direct values because e.g. the max value of `m1` depends on the given value of `m2` (and there are more constraints of that kind).

The bottom line is, it’s probably really better to dismiss the option. So unless now a professional comes along and that he really wants work with the direct values ;-) just delete the function `wagner7CustomizableDirect()` and the corresponding `exports.` line at the end of the script.

Kind regards,
Tobias

**Comment by Fil** [comment]
Thank you! I don't think D3 needs to offer every internal in its public API (and documentation) — the page on your site is here for people who will want to dig deeper.

Can you clarify that you agree to contribute this code under [D3's licence, BSD-3](https://github.com/d3/d3/blob/master/LICENSE), on top of the GPL-3 you offer on your website?

**Comment by TJung1968** [comment]
All right, then we can drop the "direct values" approach.
And yes, I fully agree to contribute the code under D3's license! :-)

**Comment by Fil** [comment]
Perfect. There was just a small mistake in the inverse (n * … should have been 1/ * …).

Please review the pull-request #145

**Comment by Fil** [comment]
Here's an observable with the build of PR #145 
https://beta.observablehq.com/d/9dd6ed94e100375a

**Comment by TJung1968** [comment]
Hmmm.
I guess you were right about the inverse after all. You see, I sometimes [Rasmus Einarsson’s raster reprojection plugin](https://bl.ocks.org/rasmuse/75fae4fee3354ec41a49d10fb37af551). Although he says it’s "not ready for production", it works fine for all projections that I’ve tried.
Now, when I try this with the inverse I wrote, the result is rubbish:
![reproject inversenotfixed](https://user-images.githubusercontent.com/7100002/44285938-7c90dd80-a267-11e8-8df5-6df7b9d42772.png)

But using `1/n * ...`, the image looks as expected:
![reproject inversefixed](https://user-images.githubusercontent.com/7100002/44285968-9af6d900-a267-11e8-9e8e-ae6c59104b1f.png)

Even considering that the plugin might be flawed, I think it’s pretty unlikely that it ny accident renders a correct projection image with faulty inverse. ;-)

**Comment by Fil** [comment]
Cool :)
Another hint that might help convince you is that with wagner7 we have
n=1/3

**Comment by TJung1968** [comment]
Wellllll… Math’s really not my strong suit.
I’m surprised myself that I even dare tinkering around with projection formulae… On the other hand, I’ve always been interested in maps. If only my math teachers back in school had said: "Now, kids, listen carefully, if you’re ever going to work with map projections, you’ll be needing this…"
I *might* have paid more attention then…
